Adds PHPStan Integration

ChrisMKindred opened the pull request:
## What does this do/fix?

Adds PHPStan
Updates the codebase to pass up to level 3 analysis.

## Brief Description

This PR pulls in PHPStan and the stubs to match packages that we are using (WordPress and ACF-Pro). We should make it a requirement that all new Pull Requests pass the PHP Static Analysis GitHub action before starting a code review. PHPStan does not take the place of code reviews or testing but helps us write more testable code and will help speed up our code reviews by being able to focus on architecture instead of formatting and typing.

### Testing Levels

The goal of using PHPStan is to get your code to the highest level of analysis.  I was able to update the code to level 3 but didn't want to take it further without getting the implementation merged.  Each level you [climb adds additional requirements](https://phpstan.org/user-guide/rule-levels).

### Patches

A composer patch runs on the stubs that are pulled in. The patch is needed because the wp_check_password(), wp_hash_pashword(), and wp_set_password() are pluggable and are overwritten using the `roots/wp-password-bcrypt`. This causes them to appear as duplicates when trying to run the code and the `php-stubs/wordpress-stubs` package.  The patch prefixes the WordPress versions of the functions to not duplicate them. Another fix applied in the patch is spreading the 2nd variable in the appy_filter function so that it doesn't throw an error when calling apply_filter with more than 2 functions. 

## Tests

Does this have tests?

- [ ] Yes
- [X] No, this doesn't need tests because it is a developer utility for analyzing code.
- [ ] No, I need help figuring out how to write the tests.

tarecord reviewed (approved):
> will help speed up our code reviews by being able to focus on architecture instead of formatting and typing.

🌮 I find myself spending a lot of time on syntax... this would be great!

ChrisMKindred commented on .github/workflows/phpstan.yml:
```diff
@@ -0,0 +1,70 @@
+          composer install --prefer-dist --no-progress --no-suggest --optimize-autoloader
+
+      - name: PHPStan
+        run: ./vendor/bin/phpstan --memory-limit=512M
```
I was able to get it to fail with the default (128M) on my local. I think we can go as high as 2GB in GHA but don't think we will need to.

tarecord replied:
I'm not too worried about it but thought I'd surface it anyway 👍🏼

ChrisMKindred commented:
I have added a few additional items from the initial PR.

* Removed the need for the `apply_filters()` patch by ignoring the specific error in the `phpstan.neon.dist` file.  Any core function or method that uses ` func_get_args()` needs to be included in the ignore due to the way WP Core handles the docblocks.
* Added a benchmark file with all the remaining errors to rule level 5. Rule level five checks to ensure that the parameters expected are the same as the parameters provided in methods/functions and is the minimum we should be allowed. 
* Adds Documentation to the docs folder for running PHPStan locally.

nickpelton reviewed (approved):
👍
